sklearn/inspection/_permutation_importance.py

```diff
@@ -56,18 +58,24 @@ def _calculate_permutation_scores(
         y = _safe_indexing(y, row_indices, axis=0)
         if sample_weight is not None:
             sample_weight = _safe_indexing(sample_weight, row_indices, axis=0)
-    else:
+    elif X_type == "polars":
+        X_permuted = X.clone()  # polars DataFrame does not have `copy`
+    else:  # pandas DataFrame or numpy ndarray
         X_permuted = X.copy()
 
     scores = []
     shuffling_idx = np.arange(X_permuted.shape[0])
     for _ in range(n_repeats):
         random_state.shuffle(shuffling_idx)
-        if hasattr(X_permuted, "iloc"):
+        if X_type == "pandas":
             col = X_permuted.iloc[shuffling_idx, col_idx]
             col.index = X_permuted.index
             X_permuted[X_permuted.columns[col_idx]] = col
-        else:
+        elif X_type == "polars":
+            X_permuted = X_permuted.with_columns(
+                X_permuted.to_series(col_idx)[shuffling_idx]
+            )
```

Comment:
this seems like renaming columns? This is also under our adapters.

Reply:
This is to shuffle a specific column with the shuffling order given by `shuffling_index`.

Reply:
That to me seems like something `_safe_indexing` should handle then?

Reply:
With `col = _safe_indexing(X, col_idx, axis=1)[shuffling_idx]` we can get the shuffled column as a pandas/polars Series, but the problem is how to assign the column back. AFAIK in pandas we do `X.iloc[:, 2] = col` or `X[2] = col`, but in `polars` we may need `X.with_columns(col)`.

I looked at `_safe_assign` and it does not support `polars` now. And with my limited knowledge about `polars` I cannot easily extend the support to `polars` because I don't know how to do the equivalent of `df.iloc[row_indices, col_indices] = arr` efficiently.